imgui/src/drag_drop.rs

```diff
@@ -0,0 +1,475 @@
+    ///
+    /// Moreover, if the `DragDropSource` has also used `Condition::Once` or similar when they sent the data,
+    /// ImGui will assume its data is still valid even after your preview, so corrupting that data could
+    /// lead to all sorts of unsafe behvaior on ImGui's side. In summary, using this function for any data
+    /// which isn't truly `Copy` or "plain old data" is difficult, and requires substantial knowledge
+    /// of the various edge cases.
+    pub unsafe fn accept_payload_unchecked(
+        &self,
+        name: &ImStr,
+        flags: DragDropFlags,
+    ) -> Option<DragDropPayload> {
+        let inner = sys::igAcceptDragDropPayload(name.as_ptr(), flags.bits() as i32);
+        if inner.is_null() {
+            None
+        } else {
+            let inner = *inner;
+
+            // @fixme: there are actually other fields on `inner` which I have shorn -- they're
+            // considered internal to imgui (such as id of who sent this), so i've left it for
+            // now this way.
+            Some(DragDropPayload {
+                data: inner.Data,
+                size: inner.DataSize,
+                preview: inner.Preview,
+                delivery: inner.Delivery,
+            })
+        }
+    }
+
+    /// Ends the current target. Ironically, this doesn't really do anything in ImGui
+    /// or in imgui-rs, but it might in the future.
+    pub fn pop(self) {
+        // omitted...exists just to run Drop.
+    }
+}
+
+impl Drop for DragDropTarget<'_> {
+    fn drop(&mut self) {
+        unsafe { sys::igEndDragDropTarget() }
+    }
+}
+
+/// An empty DragDropPayload. It has no data in it, and just includes
+/// two bools with status information.
+#[derive(Debug)]
+pub struct DragDropPayloadEmpty {
+    /// Set when [`accept_payload_empty`](Self::accept_payload_empty) was called
+    /// and mouse has been hovering the target item.
+    pub preview: bool,
+
+    /// Set when [`accept_payload_empty`](Self::accept_payload_empty) was
+    /// called and mouse button is released over the target item.
+    pub delivery: bool,
+}
+
+/// A DragDropPayload with status information and some POD, or plain old data,
+/// in it.
+#[derive(Debug)]
+pub struct DragDropPayloadPod<T> {
+    /// The kind data which was requested.
+    pub data: T,
+
+    /// Set when [`accept_payload`](Self::accept_payload) was called
+    /// and mouse has been hovering the target item.
+    pub preview: bool,
+
+    /// Set when [`accept_payload`](Self::accept_payload) was
+    /// called and mouse button is released over the target item.
+    pub delivery: bool,
+}
+
+#[derive(Debug)]
+pub struct DragDropPayload {
+    /// Data which is copied and owned by ImGui. If you have accepted the payload, you can
+    /// take ownership of the data; otherwise, view it immutably. Interacting with `data` is
+    /// very unsafe.
+    pub data: *const ffi::c_void,
+
+    /// The size of the data in bytes.
+    pub size: i32,
```

Comment:
Since this type is not a mapping of a C++ type (hopefully! It's not #[repr(C)]!) we should probably use `usize` here and just handle the conversion ourselves rather than pushing it to the user.